File: src/python/modules/multistart/greedy_based_heuristics/semi_greedy_heuristic.py
from typing import List

from modules.data import InstanceData
from modules.parameters import ParameterData
from modules.solutions import Solution

from .update_solution import update_machines_indexes, update_solution
from .choose_candidate import choose_candidate
from .init_solution import get_service_order, init_solution
from .entities_greedy_heur import InsertionData
from .check_insertion import check_insertion


def get_candidate_list_by_increase_in_comp_time(
    inst: InstanceData, sol: Solution, p_job: int, d_job: int
) -> List[InsertionData]:
    cand_list = []

    for k in inst.K:
        for p_pos in range(2, len(sol.vehicles[k])):
            for d_pos in range(p_pos, len(sol.vehicles[k])):
                check_ins_data = check_insertion(
                    sol, k, p_pos, d_pos, p_job, d_job, inst
                )
                if not check_ins_data.feasible:
                    continue

                cand = InsertionData(
                    is_feasible=True,
                    cost=check_ins_data.cost,
                    pPos=p_pos,
                    dPos=d_pos,
                    pJob=p_job,
                    dJob=d_job,
                    vehicle=k,
                    possibleMachineTravels=check_ins_data.possibleMachineTravels,
                )
                cand_list.append(cand)

    return cand_list


def semi_greedy_heuristic(inst: InstanceData, params: ParameterData) -> Solution:
    sol = init_solution(inst)
    non_serviced_reqs = get_service_order(inst, params)
    idx_req_to_serve = 0
    jumped_request = False

    while idx_req_to_serve < len(non_serviced_reqs):
        p_job = non_serviced_reqs[idx_req_to_serve]
        d_job = p_job + inst.n

        cand_list = get_candidate_list_by_increase_in_comp_time(inst, sol, p_job, d_job)
        chosen = choose_candidate(cand_list, params)

        if chosen.is_feasible:
            sol = update_solution(sol, chosen, inst)
        else:
            jumped_request = True
            break  # Jumped request means stop

        idx_req_to_serve += 1

    update_machines_indexes(sol)

    sol.is_feasible = not jumped_request

    # Optional: add post-validation here if needed
    # sol.feasible = validate_solution(inst, sol, params)

    return sol

File: src/julia/modules/Multistart/src/semiGreedyHeuristic.jl
function getCandidateListByIncreaseInCompTime(
	inst::InstanceData,
	sol::Solution,
	pJob::Int64,
	d_job::Int64,
)::Vector{InsertionData}
	candList = InsertionData[]
	for k in inst.K
		for p_pos in 2:length(sol.vehicles[k])
			for dPos in p_pos:length(sol.vehicles[k])
				check_ins_data = checkInsertion(sol, k, p_pos, dPos, pJob, d_job, inst)
				if !check_ins_data.feasible
					continue
				end

				push!(candList, InsertionData(true, check_ins_data.cost, p_pos, dPos, pJob, d_job, k, check_ins_data.possibleMachineTravels))
			end
		end
	end
	return candList
end # function getCandidateListByIncreaseInCompTime()

function semiGreedyHeuristic(inst::InstanceData, params::ParameterData)::Solution
	sol = initSolution(inst)
	nonServicedReqs = copy(getServiceOrder(inst, params))
	idxReqToServe = 1
	jumpedRequest = false
	while idxReqToServe <= length(nonServicedReqs)
		pJob = nonServicedReqs[idxReqToServe]
		d_job = pJob + inst.n

		candList = getCandidateListByIncreaseInCompTime(inst, sol, pJob, d_job)
		chosen = chooseCandidate(candList, params)

		if chosen.feasible
			sol = updateSolution(sol, chosen, inst)
		else
			jumpedRequest = true
			idxReqToServe = length(nonServicedReqs)
		end
		idxReqToServe += 1
	end

	updateMachinesIndexes(sol)

	sol.feasible = !jumpedRequest

	# if  validate_solution(inst, sol, params)
	# 	sol.feasible = true
	# else
	# 	sol.feasible = false
	# end

	return sol
end # function semiGreedyHeuristic()